nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+    // this error should already be generated by some other part of NullAway still adding a check to
+    // be safe
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      return;
+    }
+    for (int i = 0; i < lhsTypeArguments.size(); i++) {
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsLHS =
+          lhsTypeArguments.get(i).getAnnotationMirrors();
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsRHS =
+          rhsTypeArguments.get(i).getAnnotationMirrors();
+      boolean isLHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsLHS.stream(), config);
+      boolean isRHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsRHS.stream(), config);
+      if (isLHSNullableAnnotated != isRHSNullableAnnotated) {
+        String lhsAnnotation = isLHSNullableAnnotated ? "Nullable" : "NonNull";
+        String rhsAnnotation = isRHSNullableAnnotated ? "Nullable" : "NonNull";
+        invalidAssignmentInstantiationError(tree, lhsAnnotation, rhsAnnotation, state, analysis);
+        return;
+      }
+      // nested generics
+      if (lhsTypeArguments.get(i).getTypeArguments().length() > 0) {
+        compareAnnotations(lhsTypeArguments.get(i), rhsTypeArguments.get(i), tree);
+      }
+    }
+  }
+
+  /**
+   * @param tree A parameterized typed tree for which we need class type with preserved annotations.
+   * @return A Type with preserved annotations.
+   */
+  private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree) {
+    Type.ClassType type = (Type.ClassType) ASTHelpers.getType(tree);
+    Preconditions.checkNotNull(type);
+    Type nullableType = NULLABLE_TYPE_SUPPLIER.get(state);
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    List<Type> newTypeArgs = new ArrayList<>();
+    for (int i = 0; i < typeArguments.size(); i++) {
+      List<Attribute.TypeCompound> myAnnos = new ArrayList<>();
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        for (JCTree.JCAnnotation annotation : annotatedType.getAnnotations()) {
+          Attribute.Compound attribute = annotation.attribute;
+          if (attribute.toString().equals("@org.jspecify.annotations.Nullable")) {
+            myAnnos.add(
+                new Attribute.TypeCompound(
+                    nullableType, com.sun.tools.javac.util.List.nil(), null));
+          }
+        }
+      }
+      // nested generics checks
+      Type currentArgType = ASTHelpers.getType(typeArguments.get(i));
+      if (currentArgType != null && currentArgType.getTypeArguments().size() > 0) {
+        Type.ClassType nestedTyp =
+            typeWithPreservedAnnotations((ParameterizedTypeTree) typeArguments.get(i));
+        newTypeArgs.add(nestedTyp);
```

Comment:
Hrm, it looks like the `myAnnos` list is not used at all in this case.  Don't we want to do the recursive call, and then take the result and still do `cloneWithMetadata` like in the `else` case?  We should add a test that fails with the current (buggy) logic